### DISCUSSION TITLE
Accessing dependencies in background tasks

### DISCUSSION THREAD

**Discussion by banuni** [fastapi/fastapi#8502] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import BackgroundTasks, FastAPI, Depends

app = FastAPI()

def add_some_record(email_id: str, db = Depends(get_db)):
    record = Record(parent=email_id)
    db.add(record)
    db.commit()


@app.post("/send-notification/{email}")
def send_notification(email: str, background_tasks: BackgroundTasks, db = Depends(get_db)):
    email_id = db.query(Email).filter(Email.address == email).first().id
    background_tasks.add_task(add_some_record, email_id=email_id)
    return {"message": "Record added in the background"}
```

### Description

Hey there, I want to be able to access my database dependency in my background task, I tried to do it like in the example code but this method doesn't work, and I couldn't find anything in the docs about the interaction of deps and background tasks, is there a way to achieve this? should I just pass in the db session to the task and will stay alive until the task is done? it would be nice to have a best practice written somewhere

### Operating System

macOS

### FastAPI Version

0.76.0

### Python Version

3.10.0

**Comment by odiseo0** [comment]
You can't use `Depends` but you can try using a context manager to open a database connection, there are many ways.
I would use:
```py
def add_some_record(email_id: str):
    with SessionLocal() as db:
        record = Record(parent=email_id)
        db.add(record)
        db.commit()
```

**Comment by xingdongzhe** [comment]
1. You can read [dependencies-with-yield](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/)
2. You could `background_tasks.add_task(add_some_record, email_id=email_id, db=db)`

**Reply by nameer** [reply]
Wouldn't that leave the database session open for a long time? Or the database session would have closed before the code is executed, right?

**Comment by banuni** [comment]
Indeed, I ended up just passing 'DB' from the route function to the task, and it works great 😃

Regardless, does it make sense to allow background tasks to receive Dependencies ? Or is it breaking FasAPI's paradigm?

**Comment by JarroVGIT** [comment, marked as answer]
> Indeed, I ended up just passing 'DB' from the route function to the task, and it works great 😃
> 
> Regardless, does it make sense to allow background tasks to receive Dependencies ? Or is it breaking FasAPI's paradigm?

It is kinda breaking Fastapi's paradigm. Anytime you pass an `param: ParamType = Depends(some_callable)` in a path operation function, the variable `param` is actually holding a `Depends` object, holding reference to the callable. This path operation function is then handled by the preceding decorator (like `@app.get("/path")`). When your endpoint is called, the decorator first resolves the `Depends` object and then passes the outcome to the actual path operation function (in the above example, it would actually pass an object of type `ParamType`. 

So, from an Fastapi perspective, it would not make sense to have arbitrary functions (which is what are put into background tasks) be first resolved for dependencies. However, you could mimic it yourself, by creating a decorator yourself that solely resolves dependencies for any given function.

**Comment by poolpitako** [comment]
> However, you could mimic it yourself, by creating a decorator yourself that solely resolves dependencies for any given function.

How?
There is no way to grab the dependency graph and ask for a dependency?
If I have an app level dependency I don't want to be passing it through every instance constructor of my project.

**Comment by gpkc** [comment]
Keep in mind your session handling when passing something like `db` to a background task. Is `db` a session? Is it closed when the dependency ends? For example, does it look like this?

```
async def get_db():
    async with sessionmanager.session() as session:
        yield session
```

If so, I'm actually not sure how the dependency lifecycle of FastAPI will work here in terms of when is this dependency closed. Is it gonna last for the whole lifecycle of the background task?

If so, consider that your background task might mess up whatever it is that you might have pushed to the database within your view, for example if you do a `db.rollback()` within the background task, since it's the same session as the view had.

I think it's safer to simply create a new session for the background task, and have it be completely independent of whatever things you had in a context manager within the view.

**Comment by antbz** [comment]
As @gpkc mentioned, indeed the dependency is only closed when the background task finishes. Besides the issues with rollbacks, you might also run into other weird cases due to this behaviour.

For example, I was buidling an endpoint `POST /task/create` that creates a task that will run in the background for a while. I also had an endpoint `GET /task/status` that retrieved the processing status of the task. I relied on a DB object being created in my `POST` endpoint so that `GET` could retrieve its status. However, I was using `session.begin()` ([docs](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#managing-transactions)) to handle my commits. This means that the transaction is only comitted to the DB when the dependency closes.

Since the dependency was kept alive until the end of the background task, regardless of whether I passed the dependency object to the task or not, the commit was not happening until the BG task finished, and so the objects where only being inserted into the DB when the task finished, which defeated the whole point of this API design. I ended up working around this by using an explicit commit and then creating a new session inside the BG task.

I'm not sure I would call this behaviour wrong, but it certainly isn't very clear that this will happen from reading the documentation. Perhaps the [Background Tasks docs](https://fastapi.tiangolo.com/tutorial/background-tasks/) could at least mention that the main endpoint's dependencies are only closed when the task itself finishes, regardless of whether variables are passed from one context to the other.

**Comment by gpkc** [comment]
It's quite weird that the dependencies don't close when a background task is running. I agree this should at least be explicitly documented. But I can't see an use-case for having dependencies hang until the background task finish, since the response handler has already returned by then. Perhaps it's to support passing them to the background task but again that doesn't feel like good design and is probably error-prone. CC @tiangolo

**Reply by christiansicari** [reply]
> It's quite weird that the dependencies don't close when a background task is running.
@gpkc I do agree at 100%, and from my experience, it's getting closed, I had evidence with my logs. I wrapped the dependency in another contextmanager function and used it independently.

**Comment by damitti444** [comment]
i have done it using the async context manager previously. 
something like this:

```python
# task_router.py

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, status
from app.task.model import TaskBase, TaskCreate
import app.task.service as task_service
from app.database.db import get_async_session
from app.task.task_functions import process_the_task

@router.post("/")
async def create_task(
    task_in: TaskBase,
    background_tasks: BackgroundTasks,
    db_session: AsyncSession = Depends(get_async_session),
    current_user: User = Depends(current_active_user),
):

    task = TaskCreate(
        **task_in.model_dump(),
        status={
            "code": "TASK_CREATED",
            "requested_by": current_user.email",
            "progress": 0,
        },
    )
    # optionally save task record to db before processing, to keep track of status and allow polling
    # also the id is handled by db
    db_task = await task_service.create(db_session=db_session, task=task)

    await db_session.commit()
    
    more_data = {
        "x":4,
        "y":300,
        }

    background_tasks.add_task(
        process_the_task,
        db_task.id,
        more_data,
    )

    return {"id": str(db_task.id), "status": db_task.status}



# task_functions.py

import contextlib
from app.database.db import get_async_session

get_async_session_context = contextlib.asynccontextmanager(get_async_session)

async def process_the_task(
    task_id: int,
    more_data: dict,
):
    try:
        async with get_async_session_context() as db_session:

            db_task = await task_service.get(db_session=db_session, task_id=task_id)

            db_task.status["code"] = "PROCESSING"
            MutableDict.changed(db_task.status) # just for postgres jsonb column
            await db_session.commit()
            
            # do some work and update progress
            for i in range(more_data["x"]):
                # do some processing
                await asyncio.sleep(more_data["y"])
               # update progress
                db_task.status["progress"] = 100 * i / more_data["x"]
                MutableDict.changed(db_task.status)
                await db_session.commit()

            db_task.status["code"] = "DONE"
            await db_session.commit()

    except Exception as e:
        db_task.status["code"] = "ERROR"
        db_task.status["error"] = str(e)
        await db_session.commit()
        raise e



# db.py
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker, create_async_engine

SQLALCHEMY_DATABASE_URL = os.environ.get("SQLALCHEMY_DATABASE_URL")

engine = create_async_engine(SQLALCHEMY_DATABASE_URL)
async_session_maker = async_sessionmaker(engine, expire_on_commit=False)

async def get_async_session() -> AsyncGenerator[AsyncSession, None]:
    async with async_session_maker() as session:
        yield session



# task_service.py
# ... crud operations for tasks
from sqlalchemy.ext.asyncio import AsyncSession

async def create(
    *, db_session: AsyncSession, task_in: TaskCreate
) -> Task:
    """Creates a task."""

    task = Task(**task_in.model_dump())

    db_session.add(task)
    await db_session.commit()
    return task

```

**Comment by chrnmaxim** [comment]
I've noticed some slightly weird behavior when passing the yielded database session from the dependency to a background task.

- Here’s my SQLAlchemy session dependency implementation using `NullPool` as poolclass and `pool_pre_ping=False`, since I'm using Odyssey (PostgreSQL connection pooler) behind PostgreSQL.

```python
engine = create_async_engine(
    url=api_settings.DATABASE_URL,
    poolclass=NullPool,
    pool_pre_ping=False,
    echo=True if api_settings.MODE == "LOCAL" else False,
    connect_args={
        "server_settings": {"application_name": api_settings.APP_NAME}
    },
)

SessionLocal = async_sessionmaker(
    bind=engine,
    autocommit=False,
    autoflush=False,
    expire_on_commit=False,
    class_=AsyncSession,
)

# Added print() calls here just to show this in the engine logs
async def get_session() -> AsyncGenerator[AsyncSession, None]:
    async with SessionLocal() as session:
        try:
            print("Before yield")
            yield session
            print("After yield")
        except Exception as ex:
            raise ex
        finally:
            print("Code in finally block")

```
- Here’s my `GET` endpoint implementation using `BackgroundTasks`:

```python
@user_router.get(
    "/count",
    summary="Get users count",
    status_code=status.HTTP_200_OK,
    response_model=None,
    responses={status.HTTP_200_OK: {"model": UserCountSchema}},
)
async def get_user_count(
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_session),
) -> UserCountSchema:
    """Get users count."""

    user_count = await UserService.get_user_count(session=session)
    # Added as an example of passing the session yielded from the dependency
    background_tasks.add_task(
        UserService.get_user_count, session=session
    )
    return UserCountSchema(user_count=user_count)
```

 - And there are logs from the db engine:

```bash
INFO:     Application startup complete.
Before yield
2025-06-25 11:40:38,620 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2025-06-25 11:40:38,620 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-06-25 11:40:38,620 INFO sqlalchemy.engine.Engine select current_schema()
2025-06-25 11:40:38,620 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-06-25 11:40:38,620 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2025-06-25 11:40:38,620 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-06-25 11:40:38,630 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-06-25 11:40:38,632 INFO sqlalchemy.engine.Engine SELECT count(*) AS count_1
FROM users
2025-06-25 11:40:38,632 INFO sqlalchemy.engine.Engine [generated in 0.00047s] ()
After yield
Code in finally block
2025-06-25 11:40:38,638 INFO sqlalchemy.engine.Engine ROLLBACK
INFO:     127.0.0.1:56868 - "GET /api/v1/users/count HTTP/1.1" 200 OK
2025-06-25 11:40:38,682 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-06-25 11:40:38,683 INFO sqlalchemy.engine.Engine SELECT count(*) AS count_1
FROM users
2025-06-25 11:40:38,683 INFO sqlalchemy.engine.Engine [cached since 0.05035s ago] ()
```
So, if I understand this correctly, SQLAlchemy calls `self.close()` inside the `AsyncSession` context manager as it exits with implicit `rollback`, which is the default `NullPool `behaviour according to the [docs](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.pool_reset_on_return), and then the session starts a transaction again in the background task, but without rollback this time.

And it also seems that if and engine was configured with `pool_pre_ping=True`, there is another `CONNECT` call to DB in a background task:
- Logs from Logfire (notice, that there are two `connect` logs to DB: first in the path function and then in background task)
![image](https://github.com/user-attachments/assets/a70a28c8-6302-4e47-babc-d7f007b3ee33)

If I just comment the SQLAlchemy AsyncSession `__aexit__ ` calls to `self.close()` there is no rollback and also the session proceeds with the transaction it opened earlier.

```python
  async def __aexit__(self, type_: Any, value: Any, traceback: Any) -> None:
      # task = asyncio.create_task(self.close())
      # await asyncio.shield(task)
      pass
```

 - Logs with commented `__aexit__ `:
```bash
INFO:     Application startup complete.
Before yield
2025-06-25 11:47:16,647 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2025-06-25 11:47:16,647 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-06-25 11:47:16,647 INFO sqlalchemy.engine.Engine select current_schema()
2025-06-25 11:47:16,647 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-06-25 11:47:16,658 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2025-06-25 11:47:16,659 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-06-25 11:47:16,661 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-06-25 11:47:16,661 INFO sqlalchemy.engine.Engine SELECT count(*) AS count_1
FROM users
2025-06-25 11:47:16,661 INFO sqlalchemy.engine.Engine [generated in 0.00047s] ()
After yield
Code in finally block
INFO:     127.0.0.1:57180 - "GET /api/v1/users/count HTTP/1.1" 200 OK
2025-06-25 11:47:16,664 INFO sqlalchemy.engine.Engine SELECT count(*) AS count_1
FROM users
2025-06-25 11:47:16,664 INFO sqlalchemy.engine.Engine [cached since 0.006711s ago] ()
```
So, it is unclear to me, how FastAPI use the yielded session even after finally block in context manager already executed (right?).
Is it possible to maintain the same transaction across both the path function and background task when using a session yielded from a dependency?
